ddtrace/contrib/botocore/services/kinesis.py

```diff
@@ -36,56 +36,69 @@ class TraceInjectionSizeExceed(Exception):
     pass
 
 
-def inject_trace_to_kinesis_stream_data(record, span):
-    # type: (Dict[str, Any], Span) -> None
+def inject_trace_to_kinesis_stream_data(record, span, stream, inject_trace_context=True):
+    # type: (Dict[str, Any], Span, str, bool) -> None
     """
     :record: contains args for the current botocore action, Kinesis record is at index 1
     :span: the span which provides the trace context to be propagated
-    Inject trace headers into the Kinesis record's Data field in addition to the existing
+    :inject_trace_context: whether to inject DataDog trace context
+    Inject trace headers and DSM headers into the Kinesis record's Data field in addition to the existing
     data. Only possible if the existing data is JSON string or base64 encoded JSON string
     Max data size per record is 1MB (https://aws.amazon.com/kinesis/data-streams/faqs/)
     """
     if "Data" not in record:
         log.warning("Unable to inject context. The kinesis stream has no data")
         return
 
-    data = record["Data"]
-    line_break, data_obj = get_kinesis_data_object(data)
-    if data_obj is not None:
-        data_obj["_datadog"] = {}
-        HTTPPropagator.inject(span.context, data_obj["_datadog"])
-        data_json = json.dumps(data_obj)
-
-        # if original string had a line break, add it back
-        if line_break is not None:
-            data_json += line_break
-
-        # check if data size will exceed max size with headers
-        data_size = len(data_json)
-        if data_size >= MAX_KINESIS_DATA_SIZE:
-            raise TraceInjectionSizeExceed(
-                "Data including trace injection ({}) exceeds ({})".format(data_size, MAX_KINESIS_DATA_SIZE)
-            )
+    # always inject if Data Stream is enabled, otherwise only inject if distributed tracing is enabled and this is the
+    # first record in the payload
+    if (config.botocore["distributed_tracing"] and inject_trace_context) or config._data_streams_enabled:
+        data = record["Data"]
+        line_break, data_obj = get_kinesis_data_object(data)
+        if data_obj is not None:
+            data_obj["_datadog"] = {}
+
+            if config.botocore["distributed_tracing"] and inject_trace_context:
+                HTTPPropagator.inject(span.context, data_obj["_datadog"])
+
+            core.dispatch("botocore.kinesis.start", [stream, data_obj["_datadog"]])
+
+            data_json = json.dumps(data_obj)
+
+            # if original string had a line break, add it back
+            if line_break is not None:
+                data_json += line_break
 
-        record["Data"] = data_json
+            # check if data size will exceed max size with headers
+            data_size = len(data_json)
+            if data_size >= MAX_KINESIS_DATA_SIZE:
```

Comment:
This now checks both trace and DSM, right?  Do we want failure to be all or nothing, or prioritize one over the other?

Reply:
Im honestly not too sure which would be the priority in that case. But yes, this does do the check for both trace and DSM injected data size at the same time